src/realtime/websocket/websocket.gateway.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+import { OnGatewayConnection, WebSocketGateway } from '@nestjs/websockets';
+import { IncomingMessage } from 'http';
+import WebSocket from 'ws';
+
+import { ConsoleLoggerService } from '../../logger/console-logger.service';
+import { NotesService } from '../../notes/notes.service';
+import { PermissionsService } from '../../permissions/permissions.service';
+import { SessionService } from '../../session/session.service';
+import { User } from '../../users/user.entity';
+import { UsersService } from '../../users/users.service';
+import { RealtimeNoteService } from '../realtime-note/realtime-note.service';
+import { WebsocketConnection } from '../realtime-note/websocket-connection';
+import { extractNoteIdFromRequestUrl } from './utils/extract-note-id-from-request-url';
+
+/**
+ * Gateway implementing the realtime logic required for realtime note editing.
+ */
+@WebSocketGateway({ path: '/realtime' })
+export class WebsocketGateway implements OnGatewayConnection {
+  constructor(
+    private readonly logger: ConsoleLoggerService,
+    private noteService: NotesService,
+    private realtimeNoteService: RealtimeNoteService,
+    private userService: UsersService,
+    private permissionsService: PermissionsService,
+    private sessionService: SessionService,
+  ) {
+    this.logger.setContext(WebsocketGateway.name);
+  }
+
+  /**
+   * Handler that is called for each new WebSocket client connection.
+   * Checks whether the requested URL path is valid, whether the requested note
+   * exists and whether the requesting user has access to the note.
+   * Closes the connection to the client if one of the conditions does not apply.
+   *
+   * @param clientSocket The WebSocket client object.
+   * @param request The underlying HTTP request of the WebSocket connection.
+   */
+  async handleConnection(
+    clientSocket: WebSocket,
+    request: IncomingMessage,
+  ): Promise<void> {
+    try {
+      const user = await this.findUserByRequestSession(request);
+      const note = await this.noteService.getNoteByIdOrAlias(
+        extractNoteIdFromRequestUrl(request),
+      );
+
+      if (!(await this.permissionsService.mayRead(user, note))) {
+        this.logger.log(
+          `Access denied to note '${note.id}' for user '${user.username}'`,
+          'handleConnection',
+        );
```

Comment:
We had a TODO here, that we should inform the client about the failure, so that the client does not try to reconnect the closed connection.
Do we want to re-add this TODO here?